pkg/client/query/cache/memory.go

```diff
@@ -0,0 +1,378 @@
+			nearestCachedVersion = cachedVersion
+			// DEV_NOTE: Since the list is sorted in descending order, once we
+			// encounter a cachedVersion that is less than or equal to version,
+			// all subsequent cachedVersions SHOULD also be less than or equal to
+			// version.
+			break
+		}
+	}
+
+	if nearestCachedVersion == -1 {
+		return zero, ErrCacheMiss.Wrapf("key: %s, version: %d", key, version)
+	}
+
+	value, exists := valueHistory.versionToValueMap[nearestCachedVersion]
+	if !exists {
+		// DEV_NOTE: This SHOULD NEVER happen. If it does, it means that the cache has been corrupted.
+		return zero, ErrCacheInternal.Wrapf("failed to load historical value for key: %s, version: %d", key, version)
+	}
+
+	isTTLEnabled := c.config.ttl > 0
+	isCacheValueExpired := time.Since(value.cachedAt) > c.config.ttl
+	if isTTLEnabled && isCacheValueExpired {
+		// DEV_NOTE: Intentionally not pruning here to improve concurrent speed;
+		// otherwise, the read lock would be insufficient. The value will be pruned
+		// in the subsequent call to SetAsOfVersion() after c.config.maxVersionAge
+		// blocks have elapsed. If usage is such that historical values aren't being
+		// subsequently set, numHistoricalBlocks (if configured) will eventually
+		// cause the pruning of historical values with expired TTLs.
+		return zero, ErrCacheMiss.Wrapf("key: %s, version: %d", key, version)
+	}
+
+	return value.value, nil
+}
+
+// Set adds or updates the value in the cache for the given key. If the cache is
+// configured for historical mode, it will store the value at the latest **known**
+// version, which is only updated on calls to SetAsOfVersion, and therefore is not
+// guaranteed to be the current version w.r.t. the blockchain.
+func (c *inMemoryCache[T]) Set(key string, value T) error {
+	if c.config.historical {
+		return ErrUnsupportedHistoricalModeOp.Wrap("inMemoryCache#Set() is not supported in historical mode")
+	}
+
+	c.valuesMu.Lock()
+	defer c.valuesMu.Unlock()
+
+	isMaxKeysConfigured := c.config.maxKeys > 0
+	cacheMaxKeysReached := int64(len(c.values)) >= c.config.maxKeys
+	if isMaxKeysConfigured && cacheMaxKeysReached {
+		if err := c.evict(); err != nil {
+			return err
+		}
+	}
+
+	c.values[key] = cacheValue[T]{
+		value:    value,
+		cachedAt: time.Now(),
+	}
+
+	return nil
+}
+
+// SetAsOfVersion adds or updates the historical value in the cache for the given key,
+// and at the version number. If the cache is not configured for historical mode, it
+// returns an error.
+func (c *inMemoryCache[T]) SetAsOfVersion(key string, value T, version int64) error {
+	if !c.config.historical {
+		return ErrHistoricalModeNotEnabled
+	}
+
+	// Update c.latestVersion if the given version is newer (higher).
+	latestVersion := c.latestVersion.Load()
+	if version > latestVersion {
+		// NB: Only update if c.latestVersion hasn't changed since we loaded it above.
+		if c.latestVersion.CompareAndSwap(latestVersion, version) {
+			latestVersion = version
```

Comment:
```suggestion
			latestVersion = version
			c.latestVersion.Store(version)
```

Reply:
Can you elaborate on the thinking here? In this case, the `latestVersion` was updated to an even "newer/later" concurrently. The point of using `#CompareAndSwap()` here was to avoid the latest version from regressing.

Reply:
My bad. I overlooked the fact that `CompareAndSwap` do swap too :sweat_smile: